lib/chef/provider/package/rubygems.rb

```diff
@@ -482,7 +482,7 @@ def cleanup_after_converge
 
         def candidate_version
           @candidate_version ||= begin
-                                   if target_version_already_installed?(@current_resource.version, @new_resource.version)
+                                   if action != :upgrade && target_version_already_installed?(@current_resource.version, @new_resource.version)
```

Comment:
Yeah, that's the bug in the superclass I was referring to.

Its really the wrong way to go about solving this problem by lying to the superclass that there's no candidate_version.  The superclass should be smart enough to see that you're trying to `:install` and therefore if the package is installed and satisfies the resource it should never check the candidate_version.

Otherwise this check gets pushed down into *all* of the subclasses and they all have to lie to the superclass, and I'm certain that they do not.  If we want to maintain this hack then really there should be a candidate_version in the superclass that implements this check for all the subclasses, and then the subclasses should call super before adding their implementation (or the superclass should call some wrapper around candidate_version that does the lying when action is :install).